File: core/x/javaapi/api/src/main/java/viaduct/java/api/documents/MutationFromAnnotation.java
package viaduct.java.api.documents;

import viaduct.java.api.annotations.GraphQLOperation;

/** Base class for reusable, build-time validated GraphQL mutation operations. */
public abstract class MutationFromAnnotation {
  /** Returns the operation document declared by {@link GraphQLOperation}. */
  public final String getOperationText() {
    GraphQLOperation annotation = getClass().getAnnotation(GraphQLOperation.class);
    if (annotation == null) {
      throw new IllegalStateException(
          getClass().getSimpleName() + " must be annotated with @GraphQLOperation");
    }
    return annotation.value();
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/documents/MutationFromAnnotation.kt
package viaduct.api.documents

import viaduct.apiannotations.StableApi

/**
 * Abstract base class for named GraphQL mutation operation objects.
 *
 * Extend this class with a Kotlin singleton `object` and annotate it with [GraphQLOperation].
 *
 * ```kotlin
 * @GraphQLOperation("mutation { sendMessage(input: \$input) { success } }")
 * object SendMessageMutation : MutationFromAnnotation()
 * ```
 *
 * Execute via `ctx.mutation`:
 * ```kotlin
 * val result = ctx.mutation(SendMessageMutation, mapOf("input" to inputValue))
 * ```
 *
 * @see GraphQLOperation
 * @see QueryFromAnnotation
 */
@StableApi
abstract class MutationFromAnnotation {
    /**
     * The GraphQL operation document text declared in [@GraphQLOperation][GraphQLOperation].
     *
     * @throws IllegalStateException if the subclass is not annotated with [@GraphQLOperation][GraphQLOperation].
     */
    val operationText: String by lazy {
        this::class.annotations
            .filterIsInstance<GraphQLOperation>()
            .firstOrNull()
            ?.value
            ?: error("${this::class.simpleName} must be annotated with @GraphQLOperation")
    }
}
